### PR TITLE
initial push

### PR DESCRIPTION
**Pull request by AtrumAmantus**
Basic controller, some initial framework, unit tests, not too much.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        <version>2.1.6.RELEASE</version>
+        <relativePath/> <!-- lookup parent from repository -->
+    </parent>
+
+    <groupId>dev.codesupport.web</groupId>
+    <artifactId>api-service</artifactId>
+    <version>0.0.1-SNAPSHOT</version>
+    <packaging>jar</packaging>
+
+    <name>CodeSupport Api</name>
```
Comment (LamboCreeper) [inline pom.xml]: Nitpicking, but could this be `CodeSupport Website API` or `CodeSupport Website Backend`?

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    <description>Demo application showing SpringBoot REST Api and hibernate</description>
```
Comment (LamboCreeper) [inline pom.xml]: Change the description to `The backend for the CodeSupport website.`, this is no longer a demo project. :)

File: readme.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+``./mvnw install`` - Compiles, packages, and installs to local m2 repo (``<user>/.m2/repository``)
+
+``./mvnw test`` - Runs unit tests
+
+``./mvn clean package`` - Deletes `target/` folder and repackages (preferred way to build)
+
+
+##Usage
+###Build the App
```
Comment (LamboCreeper) [inline readme.md]: This entire markdown file doesn't render correctly. Can you go through and fix that please?

File: src/main/java/dev/codesupport/web/api/service/UserService.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        return userCrudOperations.createEntities(users);
+    }
+
+    public List<User> updateUsers(List<User> users) {
+        return userCrudOperations.updateEntities(users);
+    }
+
+    public List<User> deleteUsers(List<User> users) {
+        userCrudOperations.deleteEntities(users);
+        //TODO: What should this return?
```
Comment (LamboCreeper) [inline src/main/java/dev/codesupport/web/api/service/UserService.java:46]: We'll come back to this when we implement Firebase Authentication.

**Review by LamboCreeper — CHANGES_REQUESTED** [review]
See comments. :)

**Review by LamboCreeper — APPROVED** [review]
Thanks for making those changes, I'd wait until @Kayomn has reviewed too before merging due to the size of the pull request.

File: .mvn/wrapper/MavenWrapperDownloader.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.channels.Channels;
+import java.nio.channels.ReadableByteChannel;
+import java.util.Properties;
+
+public class MavenWrapperDownloader {
```
Comment (kayomn) [inline .mvn/wrapper/MavenWrapperDownloader.java:29]: Class is missing a comment?

Hard to say, there is no style guide yet.

File: .mvn/wrapper/MavenWrapperDownloader.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+     */
+    private static final String MAVEN_WRAPPER_JAR_PATH =
+            ".mvn/wrapper/maven-wrapper.jar";
+
+    /**
+     * Name of the property which should be used to override the default download url for the wrapper.
+     */
+    private static final String PROPERTY_NAME_WRAPPER_URL = "wrapperUrl";
+
+    public static void main(String args[]) {
```
Comment (kayomn) [inline .mvn/wrapper/MavenWrapperDownloader.java:55]: Again, missing comment?

What is the goal of this main procedure?

File: .mvn/wrapper/MavenWrapperDownloader.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            System.out.println("Done");
+            System.exit(0);
+        } catch (Throwable e) {
+            System.out.println("- Error downloading");
+            e.printStackTrace();
+            System.exit(1);
+        }
+    }
+
+    private static void downloadFileFromURL(String urlString, File destination) throws Exception {
```
Comment (kayomn) [inline .mvn/wrapper/MavenWrapperDownloader.java:104]: Another missing comment.

File: .mvn/wrapper/MavenWrapperDownloader.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        URL website = new URL(urlString);
+        ReadableByteChannel rbc;
+        rbc = Channels.newChannel(website.openStream());
```
Comment (kayomn) [inline .mvn/wrapper/MavenWrapperDownloader.java:107]: Could simply be:
```java
ReadableByteChannel rbc = Channels.newChannel(new URL(urlString).openStream());
```

File: mvnw [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+    CLASSPATH=`cygpath --path --unix "$CLASSPATH"`
+fi
+
+# For Mingw, ensure paths are in UNIX format before anything is touched
+if $mingw ; then
+  [ -n "$M2_HOME" ] &&
+    M2_HOME="`(cd "$M2_HOME"; pwd)`"
+  [ -n "$JAVA_HOME" ] &&
+    JAVA_HOME="`(cd "$JAVA_HOME"; pwd)`"
+  # TODO classpath?
```
Comment (kayomn) [inline mvnw:117]: Unresolved TODO in commit?

File: readme.md [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+- Common utilities
+- Error/Exception Handling
+- CRUD framework
+
+
+    dev.codesupport.web.domain.*
+
+Rich domain models (Business models) used for API interactions and business validations
+
+### Validations
```
Comment (Kayomn) [inline readme.md]: Missing section?

File: readme.md [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+Basic crud requirements are fulfilled (with preset mechanics) by utilizing the `CrudOperations` class.
+```
+getAll() - get all available entities
+getById() - get by a specific id, throws `ResourceNotFoundException`, resulting in 404, if not found
+createEntities() - saves entities to db, throws Exception if already exists by id, performs validation checks (domain & persistence level)
+updateEntities() - updates entities in db, throws `ResourceNotFoundException`, resulting in 404, if not found, performs validation checks (domain & persistence level)
+deleteEntities() - deletes entities from db, throws `ResourceNotFoundException`, resulting in 404, if not found
+```
+
+### Conventions
```
Comment (Kayomn) [inline readme.md]: Missing section?

File: src/main/java/dev/codesupport/web/api/ApiServiceApplication.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
+package dev.codesupport.web.api;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.ComponentScan;
+
+@SpringBootApplication
+@ComponentScan({"dev.codesupport"})
+public class ApiServiceApplication {
```
Comment (kayomn) [inline src/main/java/dev/codesupport/web/api/ApiServiceApplication.java:14]: Missing comment?

What does this class do?

File: src/main/java/dev/codesupport/web/api/ApiServiceApplication.java [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+    public static void main(String[] args) {
```
Comment (kayomn) [inline src/main/java/dev/codesupport/web/api/ApiServiceApplication.java:16]: Missing comment?

What does this procedure do?

File: src/main/java/dev/codesupport/web/api/config/ApplicationConfiguration.java [hunk 14]
```diff
@@ -0,0 +1,18 @@
+package dev.codesupport.web.api.config;
+
+import dev.codesupport.web.common.service.service.CrudOperations;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+class ApplicationConfiguration {
```
Comment (kayomn) [inline src/main/java/dev/codesupport/web/api/config/ApplicationConfiguration.java:12]: Missing comment?

What does this class do?

File: src/main/java/dev/codesupport/web/api/ApiServiceApplication.java (hunk 12 above)
Comment (kayomn) [inline src/main/java/dev/codesupport/web/api/ApiServiceApplication.java:14]: Missing protection level?

File: src/main/java/dev/codesupport/web/api/config/ApplicationConfiguration.java (hunk 14 above)
Comment (kayomn) [inline src/main/java/dev/codesupport/web/api/config/ApplicationConfiguration.java:12]: Missing protection level?

File: src/main/java/dev/codesupport/web/api/config/ApplicationConfiguration.java [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+
+    @Autowired
+    ApplicationConfiguration(
```
Comment (Kayomn) [inline src/main/java/dev/codesupport/web/api/config/ApplicationConfiguration.java]: What is this formatting?

File: src/main/java/dev/codesupport/web/api/config/ApplicationConfiguration.java [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+            ApplicationContext context
+    ) {
```
Comment (kayomn) [inline src/main/java/dev/codesupport/web/api/config/ApplicationConfiguration.java:24]: What is this formatting?

File: src/main/java/dev/codesupport/web/api/config/ApplicationConfiguration.java (hunk 14 above)
Comment (kayomn) [inline src/main/java/dev/codesupport/web/api/config/ApplicationConfiguration.java:12]: Why can this not just be a static class?

File: src/main/java/dev/codesupport/web/api/config/ApplicationConfiguration.java [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+        CrudOperations.setContext(context);
+    }
```
Comment (kayomn) [inline src/main/java/dev/codesupport/web/api/config/ApplicationConfiguration.java:26]: Why is a constructor being used for something that could very easily be a static procedure on utility class?

File: src/main/java/dev/codesupport/web/api/controller/UserController.java [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+
+    @PostMapping("/users")
+    RestResponse<User> createUsers(@RequestBody List<User> users);
+
+    @PutMapping("/users")
+    RestResponse<User> updateUsers(@RequestBody List<User> users);
+
+    @DeleteMapping("/users")
+    RestResponse<User> deleteUsers(@RequestBody List<User> users);
+}
```
Comment (kayomn) [inline src/main/java/dev/codesupport/web/api/controller/UserController.java:39]: Everything in here is missing comments.

File: src/main/java/dev/codesupport/web/api/controller/UserControllerImpl.java [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+    @Override
+    public RestResponse<User> updateUsers(List<User> users) {
+        return RestResponse.restResponse(service.updateUsers(users));
+    }
+
+    @Override
+    public RestResponse<User> deleteUsers(List<User> users) {
+        return RestResponse.restResponse(service.deleteUsers(users));
+    }
+}
```
Comment (kayomn) [inline src/main/java/dev/codesupport/web/api/controller/UserControllerImpl.java:52]: Everything in here is missing comments.

File: src/main/java/dev/codesupport/web/api/data/entity/UserEntity.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+    private Long joinDate;
+
+    //unused - Called by hibernate under the covers
+    @SuppressWarnings("unused")
+    @PrePersist
+    private void prePersistLogic() {
+        joinDate = System.currentTimeMillis();
+    }
+
+}
```
Comment (kayomn) [inline src/main/java/dev/codesupport/web/api/data/entity/UserEntity.java:42]: All of this is missing comments.

**Pull request merged**